modules/core/src/main/java/org/apache/ignite/internal/processors/authentication/IgniteAuthenticationProcessor.java

```diff
@@ -971,6 +893,81 @@ private static boolean isNodeHoldsUsers(ClusterNode n) {
         return !n.isClient() && !n.isDaemon();
     }
 
+    /** {@inheritDoc} */
+    @Override public SecurityContext authenticateNode(ClusterNode node, SecurityCredentials cred) throws IgniteCheckedException {
+        return new SecurityContextImpl(
+            node.id(),
+            node.attribute(ATTR_IGNITE_INSTANCE_NAME),
+            REMOTE_NODE,
+            new InetSocketAddress(F.first(node.addresses()), 0));
+    }
+
+    /** {@inheritDoc} */
+    @Override public SecuritySubject authenticatedSubject(UUID subjId) throws IgniteCheckedException {
+        throw new UnsupportedOperationException();
+    }
+
+    /** {@inheritDoc} */
+    @Override public Collection<SecuritySubject> authenticatedSubjects() throws IgniteCheckedException {
+        throw new UnsupportedOperationException();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isGlobalNodeAuthentication() {
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void authorize(String name, SecurityPermission perm, SecurityContext securityCtx) throws SecurityException {
+        if (!userOps.contains(perm))
+            return;
+
+        SecuritySubject subj = ctx.security().securityContext().subject();
+
+        Object login = subj.login();
+
+        if (subj.type() == REMOTE_NODE) {
+            throw new SecurityException("User management operations initiated on behalf of the Ignite node" +
+                " are not supported [igniteInstanceName=" + login + ']');
+        }
+
+        if (!DEFAULT_USER_NAME.equals(login) && !(ALTER_USER == perm && Objects.equals(login, name)))
+            throw new SecurityException("User management operations are not allowed for user [curUser=" + login + ']');
+
+        if (DROP_USER == perm && DEFAULT_USER_NAME.equals(name))
+            throw new SecurityException("Default user cannot be removed.");
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onSessionExpired(UUID subjId) {
+        // No-op.
+    }
+
+    /** {@inheritDoc} */
+    @Override public SecurityContext securityContext(UUID subjId) {
+        User user = users.get(subjId);
+
+        return user == null ? null : new SecurityContextImpl(subjId, user.name(), REMOTE_CLIENT, null);
+    }
+
+    /**
+     * Gets the user with the specified ID and login. It is necessary to check the login to make sure that there was
+     * no collision when calculating the user ID.
+     */
+    private User findUser(UUID subjId, String login) {
+        User user = users.get(subjId);
```

Comment:
The users variable guarded with the mux lock in other methods. Actually this lock is tricky, I'm not fully understand what is a purpose of that. Could you please should we use this lock there too?

Reply:
My opinion is - we should not use lock there. 
1. We just read from user map that is already ConcurrentMap.
2. As I see the main purpose of this mux is to to keep user operations consistent in case they should be send to a new joining node. But frankly it does not always work as expected see  https://issues.apache.org/jira/browse/IGNITE-14301

Reply:
I'm OK with that there is the ticket 14301 for discovering that.